PhysicsTools/PatAlgos/plugins/GenJetFlavourInfoPreserver.cc

```diff
@@ -0,0 +1,103 @@
+#include "SimDataFormats/JetMatching/interface/JetFlavourInfoMatching.h"
+#include "DataFormats/HepMCCandidate/interface/GenParticle.h"
+#include "DataFormats/HepMCCandidate/interface/GenParticleFwd.h"
+
+
+namespace pat {
+
+  class GenJetFlavourInfoPreserver : public edm::stream::EDProducer<> {
+  public:
+    explicit GenJetFlavourInfoPreserver(const edm::ParameterSet & iConfig);
+    virtual ~GenJetFlavourInfoPreserver() { }
+    
+    virtual void produce(edm::Event & iEvent, const edm::EventSetup & iSetup);
+    
+  private:
+    const edm::EDGetTokenT<edm::View<reco::GenJet> > GenJetsToken_;
+    const edm::EDGetTokenT<edm::View<reco::Jet> > slimmedGenJetsToken_;
+
+    const StringCutObjectSelector<reco::GenJet> cut_;
+    
+    const edm::EDGetTokenT<reco::JetFlavourInfoMatchingCollection> ak4GenJetFlavourInfosToken_;
+  };
+
+} // namespace
+
+pat::GenJetFlavourInfoPreserver::GenJetFlavourInfoPreserver(const edm::ParameterSet & iConfig) :
+    GenJetsToken_(consumes<edm::View<reco::GenJet> >(iConfig.getParameter<edm::InputTag>("GenJets"))),
+    slimmedGenJetsToken_(consumes<edm::View<reco::Jet> >(iConfig.getParameter<edm::InputTag>("slimmedGenJets"))),
+    cut_(iConfig.getParameter<std::string>("cut")),
+    ak4GenJetFlavourInfosToken_(consumes<reco::JetFlavourInfoMatchingCollection>(iConfig.getParameter<edm::InputTag>("GenJetFlavourInfos")))
+{
+    produces<reco::JetFlavourInfoMatchingCollection>();
+}
+
+void 
+pat::GenJetFlavourInfoPreserver::produce(edm::Event & iEvent, const edm::EventSetup & iSetup) {
+    using namespace edm;
+    using namespace std;
+
+    Handle<View<reco::GenJet> >      GenJets;
+    iEvent.getByToken(GenJetsToken_, GenJets);
+
+    Handle<View<reco::Jet> >      slimmedGenJets;
+    iEvent.getByToken(slimmedGenJetsToken_, slimmedGenJets);
+
+    Handle<reco::JetFlavourInfoMatchingCollection> ak4GenJetFlavourInfos;
+    iEvent.getByToken(ak4GenJetFlavourInfosToken_,ak4GenJetFlavourInfos);
+
+    auto jetFlavourInfos = std::make_unique<reco::JetFlavourInfoMatchingCollection>(reco::JetRefBaseProd(slimmedGenJets));
+
+
+    uint slimmedId = 0;
+
+	
+
+    for (View<reco::GenJet>::const_iterator it = GenJets->begin(), ed = GenJets->end(); it != ed; ++it) {
+        if (!cut_(*it)) continue;
+
+        for(reco::JetFlavourInfoMatchingCollection::const_iterator JetInfo  = ak4GenJetFlavourInfos->begin(); JetInfo != ak4GenJetFlavourInfos->end();++JetInfo){
+                
+                if((JetInfo - ak4GenJetFlavourInfos->begin()) < (it - GenJets->begin())) continue;
+                else if((JetInfo - ak4GenJetFlavourInfos->begin()) > (it - GenJets->begin())) continue;
+
+                (*jetFlavourInfos)[slimmedGenJets->refAt(slimmedId)] = reco::JetFlavourInfo(JetInfo->second.getbHadrons(), JetInfo->second.getcHadrons(), JetInfo->second.getPartons(), JetInfo->second.getLeptons(), JetInfo->second.getHadronFlavour(), JetInfo->second.getPartonFlavour());
```

Comment:
I don't think that this synchronization with slimmed GenJet indexing via a copy-paste of ```if (!cut_(*it)) continue;``` can be a working solution.
This module should just make jetFlavourInfos for all genJets and then the PATGenJetSlimmer should re-key and make a slimmed version.